src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/DoNotDefineFinalizersForTypesDerivedFromMemoryManager.cs

```diff
@@ -0,0 +1,72 @@
+    /// </summary>
+    [DiagnosticAnalyzer(LanguageNames.CSharp, LanguageNames.VisualBasic)]
+    public sealed class DoNotDefineFinalizersForTypesDerivedFromMemoryManager : DiagnosticAnalyzer
+    {
+        internal const string RuleId = "CA2015";
+
+        private static readonly LocalizableString s_localizableTitle = new LocalizableResourceString(nameof(MicrosoftNetCoreAnalyzersResources.DoNotDefineFinalizersForTypesDerivedFromMemoryManagerTitle), MicrosoftNetCoreAnalyzersResources.ResourceManager, typeof(MicrosoftNetCoreAnalyzersResources));
+        private static readonly LocalizableString s_localizableMessage = new LocalizableResourceString(nameof(MicrosoftNetCoreAnalyzersResources.DoNotDefineFinalizersForTypesDerivedFromMemoryManagerMessage), MicrosoftNetCoreAnalyzersResources.ResourceManager, typeof(MicrosoftNetCoreAnalyzersResources));
+        private static readonly LocalizableString s_localizableDescription = new LocalizableResourceString(nameof(MicrosoftNetCoreAnalyzersResources.DoNotDefineFinalizersForTypesDerivedFromMemoryManagerDescription), MicrosoftNetCoreAnalyzersResources.ResourceManager, typeof(MicrosoftNetCoreAnalyzersResources));
+
+        internal static readonly DiagnosticDescriptor Rule = DiagnosticDescriptorHelper.Create(RuleId,
+            s_localizableTitle,
+            s_localizableMessage,
+            DiagnosticCategory.Reliability,
+            RuleLevel.BuildWarning,
+            description: s_localizableDescription,
+            isPortedFxCopRule: false,
+            isDataflowRule: false);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
+
+        public override void Initialize(AnalysisContext analysisContext)
+        {
+            analysisContext.EnableConcurrentExecution();
+            analysisContext.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.Analyze);
+            analysisContext.RegisterCompilationStartAction(compilationContext =>
+            {
+                INamedTypeSymbol? attributeType = compilationContext.Compilation.GetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemAttribute);
+                if (attributeType == null)
+                {
+                    return;
+                }
+
+                compilationContext.RegisterSymbolAction(AnalyzeSymbol, SymbolKind.NamedType);
+            });
+        }
+
+        private static void AnalyzeSymbol(SymbolAnalysisContext context)
+        {
+            var namedTypeSymbol = (INamedTypeSymbol)context.Symbol;
+            var memoryManager1 = context.Compilation.GetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemBuffersMemoryManager1);
+
+            if (memoryManager1 != null && namedTypeSymbol.DerivesFromOrImplementsAnyConstructionOf(memoryManager1))
+            {
+                namedTypeSymbol.GetMembers().FirstOrDefault(m => CheckMethod(m, context));
```

Comment:
💡 This is difficult to read. LINQ is typically used for pure operations (no side effects), but in this case the call to `CheckMethod` is the method that reports diagnostics. Consider using a `foreach` loop here, or replacing `CheckMethod` with something like `TryGetFinalizer` and reporting the diagnostic here if it exists.

Reply:
I agree.  Just using a foreach loop would be easier to read, less code, and less overhead:
```C#
foreach (ISymbol symbol in namedTypeSymbol.GetMembers())
{
    if (symbol is IMethodSymbol method && method.IsFinalizer())
    {
        ... // report diagnostic
        break;
    }
}
```
If you really wanted to use LINQ, you could instead do:
```C#
ISymbol? finalizer = namedTypeSymbol.GetMembers().FirstOrDefault(m => m is IMethodSymbol method && method.IsFinalizer());
if (finalizer != null)
{
    ... // report diagnostic
}
```

Reply:
> If you really wanted to use LINQ, you could instead do:
> ISymbol? finalizer = namedTypeSymbol.GetMembers().FirstOrDefault(m => m is IMethodSymbol method && method.IsFinalizer());
> if (finalizer != null)
> {
>     ... // report diagnostic
> }

I was using that at first